internal/packages/assets.go

```diff
@@ -0,0 +1,160 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package packages
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+)
+
+// AssetType represents the type of package asset.
+type AssetType string
+
+// Supported asset types.
+const (
+	AssetTypeElasticsearchIndexTemplate  AssetType = "index_template"
+	AssetTypeElasticsearchIngestPipeline AssetType = "ingest_pipeline"
+
+	AssetTypeKibanaSavedSearch   AssetType = "search"
+	AssetTypeKibanaVisualization AssetType = "visualization"
+	AssetTypeKibanaDashboard     AssetType = "dashboard"
+	AssetTypeKibanaMap           AssetType = "map"
+)
+
+// Asset represents a package asset to be loaded into Kibana or Elasticsearch.
+type Asset struct {
+	ID         string    `json:"id"`
+	Type       AssetType `json:"type"`
+	DataStream string
+}
+
+// LoadPackageAssets parses the package contents and returns a list of assets defined by the package.
+func LoadPackageAssets(pkgRootPath string) ([]Asset, error) {
+	assets, err := loadKibanaAssets(pkgRootPath)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not load kibana assets")
+	}
+
+	a, err := loadElasticsearchAssets(pkgRootPath)
+	if err != nil {
+		return a, errors.Wrap(err, "could not load elasticsearch assets")
+	}
+	assets = append(assets, a...)
+
+	return assets, nil
+}
+
+func loadKibanaAssets(pkgRootPath string) ([]Asset, error) {
+	kibanaAssetsFolderPath := filepath.Join(pkgRootPath, "kibana")
+
+	assets, err := loadFileBasedAssets(kibanaAssetsFolderPath, AssetTypeKibanaDashboard)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not load kibana dashboard assets")
+	}
+
+	a, err := loadFileBasedAssets(kibanaAssetsFolderPath, AssetTypeKibanaVisualization)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not load kibana visualization assets")
+	}
+	assets = append(assets, a...)
+
+	a, err = loadFileBasedAssets(kibanaAssetsFolderPath, AssetTypeKibanaSavedSearch)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not load kibana saved search assets")
+	}
+	assets = append(assets, a...)
+
+	a, err = loadFileBasedAssets(kibanaAssetsFolderPath, AssetTypeKibanaMap)
```

Comment:
Thinking loud: shouldn't elastic-package discover available types from the package directory? In case we add a new type the command will still operate (open-closed principle).